### PR TITLE
Migrate our AWS::S3 wrapper to use the standard aws-s3 gem

### PR DESCRIPTION
**Pull request by philbogle**
Migrate our AWS::S3 wrapper to use the standard aws-s3 gem rather than the non-Amazon Gem which is no longer supported and incompatible with Rails 4.2.

There was no testing of this component prior to this change, so I added an integration test that verifies expected behavior.  

I made sure that the integration test passed in prod both before and after the changes to s3.rb so the new behavior should be the same as the old.

### CODE REVIEW

**Comment by laurelfan** [comment]
did you look at the travis failure:

````
TypeError: superclass mismatch for class NoSuchKey
/home/travis/build/code-dot-org/code-dot-org/vendor/bundle/ruby/2.0.0/gems/aws-s3-0.6.3/lib/aws/s3/exceptions.rb:28:in `<module:S3>'
````

File: lib/cdo/aws/s3.rb [hunk 1]
```diff
@@ -1,29 +1,55 @@
...
 
 module AWS
   module S3
-    # Connects to the official AWS SDK for Ruby v2
+
+    # Region for storing S3 buckets. TODO: Move this CDO configuration.
+    S3_REGION = 'us-east-1'
+
+    # An exception class used to wrap the underlying Amazon NoSuchKey exception.
+    class NoSuchKey < Exception
```
Comment (laurelfan) [inline lib/cdo/aws/s3.rb]: why is this necessary?

File: lib/cdo/aws/s3.rb [hunk 2]
```diff
@@ -1,29 +1,55 @@
...
 
-    def self.download_from_bucket(bucket, name, options={})
-      connect!
-      AWS::S3::S3Object.find(name, bucket).value
+    # Returns the value of the specified S3 key in bucket.
+    # @param [String] bucket
+    # @param [String] key
+    # @return [String]
+    def self.download_from_bucket(bucket, key, options={})
+      create_client.get_object(bucket: bucket, key: key).body.string
```
Comment (laurelfan) [inline lib/cdo/aws/s3.rb]: where is the `.string` method from? will this work for non-text data (eg. images)?

Comment (philbogle) [inline lib/cdo/aws/s3.rb]: Fixed and added tests to verify that arbitrary byte sequences round-trip correctly.

(Both the .string and the .read method were defaulting to utf-8 encoding for the bytes read from the bucket, so it was necessary to force the encoding to BINARY.)

File: shared/test/test_aws_s3_integration.rb [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+  # An integration test of the AWS S3 wrapper that runs against the actual AWS service.
+  def test_aws_s3
+    # Test upload_to_bucket and download_from_bucket with :no_random.
+    test_key = Random.rand.to_s
+    test_value = 'hello\x00\x01\xFF'
+    upload_key = AWS::S3::upload_to_bucket(TEST_BUCKET, test_key, test_value, :no_random => true)
+    assert_equal test_value, AWS::S3::download_from_bucket(TEST_BUCKET, test_key)
+    assert_equal test_key, upload_key
+
```
Comment (philbogle) [inline shared/test/test_aws_s3_integration.rb]: fixed.

**Comment by philbogle** [comment]
Laurel, this is ready for another look. Thanks.

**Comment by laurelfan** [comment]
looks good, after resolving the travis failure

**Comment by philbogle** [comment]
Fixed the Travis failure (hopefully) by removing aws-s3 from the Gemfile, which was causing conflicting definitions for Aws::S3::NoSuchKey.

**Comment by philbogle** [comment]
Travis is happy now.
